### PR TITLE
FileWatcher_linux: Don't give up on a directory if a sub-entry couldn't be watched

### PR DESCRIPTION
**Pull request by RoaldFre**
## What does this PR do?

The file watcher on Linux used to give up on an entire directory if one of its sub-entries couldn't be watched (e.g. due to a permission problem). This PR modifies the behaviour such that the file watcher continues to (try to) add the remaining sub-entries of that directory.

## How was this PR tested?

Tested manually by adding some files to a watched folder with non-read permission.

### CODE REVIEW

File: Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp [hunk 1]
```diff
@@ -98,7 +98,7 @@ void FileWatcher::PlatformImplementation::AddWatchFolder(QString folder, bool re
             [[maybe_unused]] const auto err = errno;
             [[maybe_unused]] AZStd::fixed_string<255> errorString;
             AZ_Warning("FileWatcher", false, "inotify_add_watch failed for path %s: %s", dirName.toUtf8().constData(), strerror_r(err, errorString.data(), errorString.capacity()));
-            return;
+            continue;
```
Comment (lumberyard-employee-dm) [inline Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp:101]: Seems like we should continue only if errno matches [EACCES](https://man7.org/linux/man-pages/man2/inotify_add_watch.2.html) which indicates a permission problem

**Comment by RoaldFre** [comment]
I have improved the error handling and reporting further as per feedback (including a hint to the user for the ENOSPC case), factored out repeated code and updated the formatting of the touched parts following the current style guidelines.

The previous CI failure happened in the Windows test suite (note that this PR should only affect the Linux side): the Gem::Atom_RPI.Tests.main failure seems related to https://github.com/o3de/o3de/pull/11269; the segfault in AZ::AzCore.Tests.main seems related to https://github.com/o3de/o3de/issues/11238 . I've merged the dev branch to get the fix for the first issue.

**Review by spham-amzn — APPROVED** [review]

File: Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp [hunk 2]
```diff
@@ -63,46 +63,62 @@ void FileWatcher::PlatformImplementation::AddWatchFolder(QString folder, bool re
...
+            case EACCES:
+            case EBADF:
+            case ENOENT:
+                // Errors specific to the file: try next one
+                continue;
+            default:
+                // Other errors are usually non-recoverable: bail out to avoid warning spam
+                AZ_Warning(
+                    "FileWatcher", false,
+                    "Giving up on directory %s due to errors", cleanPath.toUtf8().constData());
```
Comment (spham-amzn) [inline Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp:118]: maybe helpful to capture the errno value in the warning message

**Review by AMZN-stankowi — APPROVED** [review]
After addressing the other feedback, especially the improved error message, this looks good to go. It's lacking automated tests, I've cut a ticket to our backlog to add those later, unless you feel comfortable adding them yourself soon.

**Review by byrcolin — APPROVED** [review]

**Comment by lmbr-pip** [comment]
Looked this failed its first AR run: https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-11248/1/ due to a SEGFault and an ATOM failure on windows

Am trying AR run again as these are just Linux changes but @RoaldFre we may need to merge in latest to get around these AR failures.

File: Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp [hunk 3]
```diff
@@ -63,46 +63,62 @@ void FileWatcher::PlatformImplementation::AddWatchFolder(QString folder, bool re
         return;
     }
 
+    auto tryToWatch = [&](const QString& pathStr) -> bool {
```
[the remark this answers is not in the document]
Comment (RoaldFre) [reply Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp]: No deep reason. I've now split it off to a method which makes the errno handling a bit more explicit.

**Comment by RoaldFre** [comment]
I've upgraded the TryToWatch() function from a lambda to a proper method (with more explicit handling of the errno), and the errno is now directly reported in warnings as well as per feedback.

I'll leave the automated tests to someone more familiar with the testing framework if possible ;-).

The latest AR (before this last commit) seemed successful so I've left the branch there to avoid another (spammy?) dev branch merge commit for now.

**Comment by lmbr-pip** [comment]
Interestingly this passed on AR: https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-11248/ but is not approved for build. Need to follow up with testing team

**Comment by amzn-changml** [comment]
This is good to go, as it passed validation previously, but GitHub Actions is failing to fire. We can bypass this as soon as we get word that we're ready to merge

File: Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp [hunk 4]
```diff
@@ -56,40 +56,45 @@ void FileWatcher::PlatformImplementation::Finalize()
     m_inotifyHandle = -1;
 }
 
+bool FileWatcher::PlatformImplementation::TryToWatch(const QString &pathStr, int* errnoPtr)
+{
+    const char* path = pathStr.toUtf8().constData();
```
Comment (lumberyard-employee-dm) [inline Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp:61]: This is a dangling pointer here.

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

**Comment by lumberyard-employee-dm** [comment]
> This is good to go, as it passed validation previously, but GitHub Actions is failing to fire. We can bypass this as soon as we get word that we're ready to merge

This isn't ready to merge yet.
There is major issue with referencing invalid memory.

**Comment by RoaldFre** [comment]
Good catch! That dangling pointer has now been fixed as suggested.

**Review by lumberyard-employee-dm — APPROVED** [review]
Thanks for the address the dangling pointer issue

**Pull request merged**